proposals/0023-scalar-constant-buffer-layout.md

```diff
@@ -0,0 +1,300 @@
+The addition of this feature to HLSL will not have any effect on existing HLSL
+code as a result.
+
+
+#### Layouts
+
+`legacy` layout uses the current packing rules described here:
+https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-packing-rules.
+
+`scalar` layout modifies the alignment rules for constant buffers, requiring
+that each element is packed to align with it's scalar size.
+For composite types such as vectors and matrices this requires that each
+component be aligned individually to it's size, rather than the composite type
+be aligned to it's total size.
+The intention of these rules is to match the `scalarBlockLayout` in Vulkan,
+providing easy compatibility between DXIL and SPIR-V layouts.
+
+
+### Interchange Format Additions
+
+This feature does not require any change to the DXIL language to represent
+scalar layouts.
+
+SPIR-V already has full support for scalar layouts.
+
+
+### Diagnostic Changes
+
+These changes introduce two new errors for the `bufferlayout` attribute.
+
+First if the `bufferlayout` attribute is applied to any definition other than a
+constant buffer an error must be reported.
+
+Second if the `bufferlayout` attribute is given an invalid layout name an error
+must be reported.
+
+
+### Validation Changes
+
+DXIL Validation must be updated to accept scalar layout constant buffers in
+addition to the existing layout constant buffers.
+
+
+## Examples
+
+The following example demonstrates legal uses of the attribute.
+
+```
+[bufferlayout("legacy")]
+cbuffer MyBuffer ...
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer ...
+
+[bufferlayout("legacy")]
+ConstantBuffer<MyStruct> MyBuffer ...
+
+[bufferlayout("scalar")]
+ConstantBuffer<MyStruct> MyBuffer ...
+```
+
+The following example demonstrates the difference in buffer alignment and size
+between legacy and scalar layouts in DXIL.
+Note that in legacy layout the float3 is aligned such that it begins in the
+second 16 byte space, while in scalar layout it begins immediately after the
+float2, with the 3rd element starting at offset 16.
+
+```
+[bufferlayout("legacy")]
+cbuffer MyBuffer {
+  float2 Foo;
+  float3 Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float2 Foo;                                   ; Offset:    0
+;       float3 Bar;                                   ; Offset:   16
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    28
+;
+; }
+
+[bufferlayout("scalar")]
+cbuffer MyBuffer {
+  float2 Foo;
+  float3 Bar;
+};
+
+; cbuffer MyBuffer
+; {
+;
+;   struct MyBuffer
+;   {
+;
+;       float2 Foo;                                   ; Offset:    0
+;       float3 Bar;                                   ; Offset:    8
+;
+;   } MyBuffer;                                       ; Offset:    0 Size:    20
+;
+; }
+```
+
+As all values must be aligned to their scalar size, a scalar or element of a
+composite type may not lay across the 16 byte boundary.
```

Comment:
Is this supposed to say 4-byte? Or was this meant to refer to the legacy layout?

Reply:
I think the language is probably a bit redundant. If scalar objects are aligned to their size for all scalar objects under 16-bytes, it would be impossible to cross a 16-byte boundary. HLSL doesn't have any larger than 8-bytes, but I don't think we need language about not crossing a 4-byte boundary because 64-bit types would be 8-byte aligned and will cross the 4-byte boundary between 8-byte alignments.